### ISSUE TITLE
Package / library identifiers

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#4029]
If I want, say, to use to different versions of some module, say `Data.Stream`, from two different external libraries, say `standard-library` and `foo`, I think there is no nice way in Agda to do that.

A most basic way to fix that is to introduce _package prefixes_ into module names for disambiguation, e.g.
```agda
import standard-library:Data.Stream
import foo:Data.Stream
```
Naively, this will fail still if both `Data.Stream` modules import a module that is present in both libraries, say `Data.Thunk`.

More robust seems to allow to add prefixes to module trees in the project (`.agda-lib`) file:
```
include:
  . as root
  ../examples as examples
  ../test as tests
depend:
  standard-library as std-lib
  foo as streams
```
These prefixes could be used with ordinary module syntax:
```agda
import std-lib.Data.Stream
import streams.Data.Stream
```
Further, imports within module trees should be resolved within the context of that tree.  This isn't the case now.  For instance, if I define locally a module `Data.Bool.Base` and try to import `Data.Bool` from the standard library, it complains about an ambiguous module `Data.Bool.Base`.  This is because we use _dynamic binding_ of file modules and mixing when we should use _static binding_.

Appropos mixing:  When we put the following in our `.agda-lib` file:
```
depend:
  foo
  bar
```
then packages can interdepend, e.g. `foo` can import modules from `bar` and vice versa.  This _open package_ approach could be called a feature, but I'd call the lack of static scope for imports a deficiency; it does not fit the spirit of Agda.  A package should be self-contained and only depend on previously defined things, in the same way that modules do.

**Comment by nad** [comment]
You take up several things at once:
* I definitely think that "A package should be self-contained and only depend on previously defined things, in the same way that modules do". See also #2770.
* How to handle duplicate module names is a different question. There is a GHC extension called PackageImports which is similar to one of your suggestions. However, I've rarely seen it used in practice, and its [documentation](https://downloads.haskell.org/~ghc/8.6.5/docs/html/users_guide/glasgow_exts.html#package-qualified-imports) states that "You probably don’t need to use this feature". GHC also provides a related feature that may be of interest: [thinning and renaming](https://downloads.haskell.org/~ghc/8.6.5/docs/html/users_guide/packages.html#package-thinning-and-renaming).

**Comment by jespercockx** [comment]
@andreasabel why is this labeled as a bug rather than an enhancement?

**Comment by jespercockx** [comment]
> How to handle duplicate module names is a different question. There is a GHC extension called PackageImports which is similar to one of your suggestions. However, I've rarely seen it used in practice, and its documentation states that "You probably don’t need to use this feature". GHC also provides a related feature that may be of interest: thinning and renaming.

Maybe a better idea is to allow `using`, `renaming`, and `hiding` statements to appear in the list of package dependencies. For example:

```
depend:
  std-lib hiding (Data.Stream)
  foo
```
or
```
depend:
  std-lib
  foo renaming (Data.Stream to Data.Stream.Foo)
```
if you want to use both.

(see also https://downloads.haskell.org/~ghc/8.6.5/docs/html/users_guide/packages.html#package-thinning-and-renaming)

**Comment by pnlph** [comment]
Sorry for the long answer. I stumbled upon this thread from #4708 

I second the solution of adding prefixes to module trees in the project `.agda-lib` file:
```
include:
  . as root
  ../examples as examples
  ../test as tests
depend:
  standard-library as std-lib
  foo as streams
```

However, I also think that the standard libraries (at least the ones who live in the agda community) should be reserved, eg. `Agda`, `Standard`, `Cubical`.

Maybe it would be useful to think about what the end-user should code (a kind of Rückwärtsterminierung) to choose an implementation.

```agda
-- no libraries loaded
module foo where                    -- "foo" distinct from "reflection"
open import Agda.Builtin.Reflection
```

Now loading a library:
```agda
-- agda-stdlib loaded (current)
module foo where                    -- "foo" distinct from "reflection"
open import Agda.Builtin.Reflection
open import Reflection
```

I would also allow (without the user having to make any changes in the `agda-lib` file) because just reading the code is enough to know to which `Reflection` the name `Standard.Reflection` refers, so the reader does not have to know the configuration in `agda-lib` to understand the code:
```agda
-- agda-stdlib loaded (additional proposal)
module foo where                    -- "foo" distinct from "reflection"
open import Agda.Builtin.Reflection
open import Standard.Reflection
```

Now, supposing that cubical would have a `Reflection`module:
```agda
-- more than one library loaded with the same module name
module foo where                    -- "foo" distinct from "reflection"
open import Agda.Builtin.Reflection
open import Standard.Reflection
open import Cubical.Reflection
```

Finally, the user wants to work an own `Reflection` module, possibly using names from other `Reflection` modules:
```agda
-- more than one library loaded with the same module name
-- module name is the same as imported modules
module reflection where
open import Agda.Builtin.Reflection
open import Standard.Reflection
open import Cubical.Reflection
```

Now, what should happen when another module wants to import the user's `Reflection` module?
```agda
-- more than one library loaded with the same module name
-- the same module name on ROOT
module tactics where
open import Reflection
open import Agda.Builtin.Reflection
open import Standard.Reflection
open import Cubical.Reflection
```

We could reserve additionally something like `ROOT` (to refer the current project) to achieve more clarity in the code:
```agda
-- more than one library loaded with the same module name
-- the same module name on ROOT
module tactics where
open import ROOT.Reflection
open import Agda.Builtin.Reflection
open import Standard.Reflection
open import Cubical.Reflection
```

In case of conflicts with names of terms, the user could easily employ the modifiers `using` and `renaming`.

**Comment by guillaumebrunerie** [comment]
I don’t see any reason for having reserved words. What if you want to call one of your files `standard.agda`? And how is `open import Standard.Reflection` better than `open import std-lib:Reflection`?

**Comment by pnlph** [comment]
> And how is `open import Standard.Reflection` better than `open import std-lib:Reflection`?

You are right. The connection between `standard-library` and `std-lib` is renamed in the `agda-lib` file, and therefore is invisible for the reader; it could have also been renamed as `pinky`. If so, the declaration `open import pinky:Reflection` only makes sense when you know what `pinky` means, but a reader has no chance of knowing that by just looking at the code.

Maybe there should be a way for the reader of the code to know which library is behind that name, would it be possible to declare this renaming within the agda code?

Otherwise, it results in the impossibility of posting, for instance, self-contained code, like happened to me in #4708 . It was vital to know that the `agda-stdlib` was loaded.

About the reserved words, for example haskell packages names are also reserved, or am I missing something here?

**Comment by pnlph** [comment]
> What if you want to call one of your files `standard.agda`?

I also think that a user should be able to call a file `standard.agda`, but when importing it from another file, maybe one should use something like:
```agda
open import ROOT.standard
```
or
```agda
open import .standard
```
I mean, anything that makes clear that it is a module defined "locally".
